tools/exporters/data_dictionary_builder.php

```diff
@@ -0,0 +1,361 @@
+<?php
+/**
+ * Purpose:
+ * This script automatically generates the following tables from ip_output.txt
+ *    - parameter_type
+ *    - parameter_type_category
+ *    - parameter_type_category_rel
+ *
+ * These tables (now referred to as the parameter_type.* tables) are the data
+ * dictionary for Loris.
+ *
+ * Input:
+ * data_dictionary_builder.php takes as input the ip_output.txt file (generated
+ * by quickform_parser.php) and inserts records for each field of each
+ * discovered NDB_BVL_Instrument. To be complete, this tool must be run on an
+ * ip_output.txt file that was constructed from all instruments.
+ *
+ * *
+ * * WARNING THIS FILE ACTIVELY AFFECTS THE DATABASE INDICATED BY CONFIG.XML.
+ * * READ THE SCRIPT FIRST AND MODIFY AS NECESSARY.
+ * *
+ *
+ * Functioning:
+ * Currently, this script deletes all entries in parameter_type.* relating to
+ * parameter_type_category.Type='Instrument' and regenerates them.
+ * parameter_type primary keys (ParameterTypeID) are preserved though
+ * parameter_type_category primary keys ARE NOT preserved.  Entries relating
+ * to parameter_type_category.Type Metafields or manually entered custom fields
+ * of parameter_type_category.Type != 'Instrument' are preserved.
+ *
+ * Further Improvements:
+ * It is recommended that this script be changed to be intelligent about its
+ * methods, possibly with command line options that allow it to be more
+ * selective about its actions and deletion targets.
+ *
+ * Usage:
+ *    php data_dictionary_builder.php
+ *
+ * PHP Version 7
+ *
+ * @category Behavioural
+ * @package  Main
+ * @author   Loris Team <loris.mni@bic.mni.mcgill.ca>
+ * @license  Loris License
+ * @link     https://github.com/aces/Loris
+ */
+require_once __DIR__ . '/../generic_includes.php';
+
+$parameter_types = array();
+$instrumentParameterTypeCategoryIDs = array();
+$instrumentParameterTypeIDs         = array();
+
+
+$parameter_types = $DB->pselectColWithIndexKey(
+    "Select Name, ParameterTypeID from parameter_type",
+    array(),
+    "Name"
+);
```

Comment:
old code:
```
getColumns(	
    "Select Name, ParameterTypeID from parameter_type",	
    $DB,	
    $parameter_types,	
    array()	
);
...
function getColumns($query, &$DB, &$result, $preparray)	
{	
    $TwoDArray = $DB->pselect($query, $preparray);	
    foreach ($TwoDArray as $containers=>$cells) {	
        $values = array_values($cells);	
        $result[$values[0]] = $values[1];	
    }	
    return $result;	
}
```

The function used here `pselectColWithIndexKey()` throws an error when it gets 2 rows with the same `Name`. The error is there to prevent data from being overwritten. Currently, all old projects will get an error at this stage because the `parameter_type` table was not well maintained.

when looking at the old code we see that it was doing the something similarto `pselectColWithIndexKey()` but it allows for a non-unique `name` value (it just overwrites the old value with the new one)... the problem is that this `$parameter_types` array result is used to prevent existing parameter type IDs from changing accross time and given that the old code allowed for duplicates on `Name` there is no guarantee that the "saved" parameter type ID is a valid one, it's just the last one in the list of fields with the same `Name`

All this to say, this error SHOULD be thrown but we might need to find a workaround for current projects to allow for some time to clean their data before they just get errors on this script

Reply:
@ridz1208 There is no garante that the ordering will the earliest to latest in your query.

Change to "Select Name, ParameterTypeID from parameter_type order by ParameterTypeID desc limit 1" will get you the latest entry in the table when multiple row have the same name.

Reply:
ordering is not important

Reply:
Using my query (leaving the `order by ParameterTypeID desc` part) would still be more efficient than you multi-lines php code. Just add the `limit 1` to your query.